### PR TITLE
Refactor state handling

### PR DESCRIPTION
**Pull request by nicodh**
Implements new state handling as proposed in #785

### CODE REVIEW

File: src/renderer/stores/locations.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+ipcRenderer.on('DD_EVENT_LOCATIONS_UPDATED', (evt, payload) => {
+  const { locations } = payload
+  locationStore.setState({ ...locationStore.getState(), locations: locations })
+})
+
+locationStore.reducers.push((action, state) => {
+  if (action.type === 'DC_GET_LOCATIONS') {
+    const { timestampFrom, timestampTo } = action.payload
+    state = Object.assign({}, state, { mapSettings: { timestampFrom: timestampFrom, timestampTo: timestampTo } })
```
Comment (ralphtheninja) [inline src/renderer/stores/locations.js]: `state = Object.assign({}, state, { mapSettings: { timestampFrom, timestampTo } })`

File: src/renderer/stores/locations.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  const { selectedChat, mapSettings } = locationStore.getState()
+  const { chatId } = payload
+  if (selectedChat && chatId === selectedChat.id) {
+    getLocations(chatId, mapSettings)
+  }
+})
+
+ipcRenderer.on('DD_EVENT_LOCATIONS_UPDATED', (evt, payload) => {
+  const { locations } = payload
+  locationStore.setState({ ...locationStore.getState(), locations: locations })
```
Comment (ralphtheninja) [inline src/renderer/stores/locations.js]: ` locationStore.setState({ ...locationStore.getState(), locations })`

File: src/renderer/stores/locations.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+})
+
+ipcRenderer.on('DC_EVENT_INCOMING_MSG', (evt, payload) => {
+  const { selectedChat, mapSettings } = locationStore.getState()
+  const { chatId } = payload
+  if (selectedChat && chatId === selectedChat.id) {
+    getLocations(chatId, mapSettings)
+  }
+})
```
Comment (ralphtheninja) [inline src/renderer/stores/locations.js]: Identical functions for `DC_EVENT_MSGS_CHANGED` and `DC_EVENT_INCOMING_MSG`.

File: src/renderer/stores/store.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    this.state = state
+    this.listeners.forEach(listener => listener(this.state))
+  }
+
+  subscribe (listener) {
+    this.listeners.push(listener)
+    return () => {
+      const index = this.listeners.indexOf(listener)
+      this.listeners.splice(index, 1)
+    }
```
Comment (ralphtheninja) [inline src/renderer/stores/store.js]: Maybe do `return this.unsubscribe.bind(this, listener)`?

File: src/renderer/stores/locations.js [hunk 5]
```diff
@@ -39,31 +47,14 @@ ipcRenderer.on('DC_EVENT_LOCATION_CHANGED', (evt, payload) => {
...
-  }
-})
-
-ipcRenderer.on('DD_EVENT_LOCATIONS_UPDATED', (evt, payload) => {
-  const { locations } = payload
-  locationStore.setState({ ...locationStore.getState(), locations: locations })
-})
+ipcRenderer.on('DC_EVENT_MSGS_CHANGED', onLocationChange)
+ipcRenderer.on('DC_EVENT_INCOMING_MSG', onLocationChange)
+ipcRenderer.on('DD_EVENT_LOCATIONS_UPDATED', onLocationChange)
```
Comment (ralphtheninja) [inline src/renderer/stores/locations.js]: This seems a bit odd. The callback doesn't look the same for this event and is `DD_EVENT_LOCATIONS_UPDATED` correct?

Comment (ralphtheninja) [reply src/renderer/stores/locations.js]: Nevermind, I thought `DD_EVENT_..` was an event from core.

Comment (nicodh) [reply src/renderer/stores/locations.js]: You were right - it was the wrong callback for that event. DD_EVENT means it's an event triggered by DeltachatDesktop

**Pull request merged**